dsc/utilities/aws/sqs.py

```diff
@@ -0,0 +1,203 @@
+
+        Args:
+            sqs_message: An SQS result message to be processed.
+        """
+        self.validate_message(sqs_message)
+        identifier = sqs_message["MessageAttributes"]["PackageID"]["StringValue"]
+        message_body = json.loads(str(sqs_message["Body"]))
+        self.delete(sqs_message["ReceiptHandle"])
+        return identifier, message_body
+
+    def receive(self) -> Iterator[MessageTypeDef]:
+        """Receive messages from SQS queue."""
+        logger.debug(f"Receiving messages from SQS queue: {self.queue_name}")
+        while True:
+            response = self.client.receive_message(
+                QueueUrl=self.queue_url,
+                MaxNumberOfMessages=10,
+                MessageAttributeNames=["All"],
+            )
+            if "Messages" in response:
+                for message in response["Messages"]:
+                    logger.debug(
+                        f"Message retrieved from SQS queue {self.queue_name}: {message}"
+                    )
+                    yield message
+            else:
+                logger.debug(f"No more messages from SQS queue: {self.queue_name}")
+                break
+
+    def send(
+        self,
+        message_attributes: Mapping[str, MessageAttributeValueTypeDef],
+        message_body: str,
+    ) -> SendMessageResultTypeDef:
+        """Send message via SQS.
+
+        Args:
+            message_attributes: The attributes of the message to send.
+            message_body: The body of the message to send.
+        """
+        logger.debug(f"Sending message to SQS queue: {self.queue_name}")
+        response = self.client.send_message(
+            QueueUrl=self.queue_url,
+            MessageAttributes=message_attributes,
+            MessageBody=str(message_body),
+        )
+        logger.debug(f"Response from SQS queue: {response}")
+        return response
+
+    def validate_message(self, sqs_message: MessageTypeDef) -> None:
+        """Validate that an SQS message is formatted as expected.
+
+        Args:
+            sqs_message:  An SQS message to be evaluated.
+
+        """
+        if not sqs_message.get("ReceiptHandle"):
+            message = f"Failed to retrieve 'ReceiptHandle' from message: {sqs_message}"
+            raise InvalidSQSMessageError(message)
+        self.validate_message_attributes(sqs_message=sqs_message)
+        self.validate_message_body(sqs_message=sqs_message)
+
+    @staticmethod
+    def validate_message_attributes(sqs_message: MessageTypeDef) -> None:
+        """Validate that "MessageAttributes" field is formatted as expected.
+
+        Args:
+            sqs_message: An SQS message to be evaluated.
+        """
+        if (
+            "MessageAttributes" not in sqs_message
+            or not any(
+                field
+                for field in sqs_message["MessageAttributes"]
+                if "PackageID" in field
+            )
```

Comment:
I must admint, this one is still tripping me up.  Are we looking for `"PackageID"` as a key in the `sqs_message["MessageAttributes"]` dictionary?  

If so, could we rewrite?

```python
or "PackageID" not in sqs_message["MessageAttributes"]
```

Reply:
Now that we are looking at the validation method even closer, I am curious about this as well. Debugging the test, it seems the an `sqs_message` looks like:

```text
{
    "MessageId": "47179327-6a75-467d-b431-d7bfd802b15e",
    "ReceiptHandle": "zqaejnmrlfkzlxsxexmzsnyjsiugkotjdouweeephwuigifrlsrwxshfqmguyjttklxxrogfiohgwezhqpywvrbgbueebxlmuenrwtgtgmhbaugtywthkppbixxbjowmkqcwiqapxjpirgeeuowpgdkrdskhawgimaaynsbekemlfxoapaskuewaa",
    "MD5OfBody": "2affecad36e208b6be933fce09b7a89e",
    "Body": '{"ResultType": "success", "ItemHandle": "1721.1/131022", "lastModified": "Thu Sep 09 17:56:39 UTC 2021", "Bitstreams": [{"BitstreamName": "10.1002-term.3131.pdf", "BitstreamUUID": "a1b2c3d4e5", "BitstreamChecksum": {"value": "a4e0f4930dfaff904fa3c6c85b0b8ecc", "checkSumAlgorithm": "MD5"}}]}',
    "MD5OfMessageAttributes": "5edd88184e055125bdab8e1b116ecf5e",
    "MessageAttributes": {
        "PackageID": {"StringValue": "10.1002/term.3131", "DataType": "String"},
        "SubmissionSource": {"StringValue": "Submission system", "DataType": "String"},
    },
}
```

Reply:
Don't know how I came up with that check and how it survived this long without refactoring but yes `or "PackageID" not in sqs_message["MessageAttributes"]` is much simpler